File: src/cpp/cdm/patient/SEMeal.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/patient/SEMeal.h"
#include "cdm/properties/SEScalarTime.h"
#include "cdm/io/protobuf/PBPatientNutrition.h"

SEMeal::SEMeal(Logger* logger) : SENutrition(logger)
{
  m_ElapsedTime = nullptr;
}

SEMeal::~SEMeal()
{
  SAFE_DELETE(m_ElapsedTime);
}

void SEMeal::Clear()
{
  SENutrition::Clear();
  INVALIDATE_PROPERTY(m_ElapsedTime);
}

bool SEMeal::SerializeToString(std::string& output, eSerializationFormat m) const
{
  return PBPatientNutrition::SerializeToString(*this, output, m);
}
bool SEMeal::SerializeToFile(const std::string& filename) const
{
  return PBPatientNutrition::SerializeToFile(*this, filename);
}
bool SEMeal::SerializeFromString(const std::string& src, eSerializationFormat m)
{
  return PBPatientNutrition::SerializeFromString(src, *this, m);
}
bool SEMeal::SerializeFromFile(const std::string& filename)
{
  return PBPatientNutrition::SerializeFromFile(filename, *this);
}

bool SEMeal::HasElapsedTime() const
{
  return m_ElapsedTime == nullptr ? false : m_ElapsedTime->IsValid();
}
SEScalarTime& SEMeal::GetElapsedTime()
{
  if (m_ElapsedTime == nullptr)
    m_ElapsedTime = new SEScalarTime();
  return *m_ElapsedTime;
}
double SEMeal::GetElapsedTime(const TimeUnit& unit) const
{
  if (m_ElapsedTime == nullptr)
    return SEScalar::dNaN();
  return m_ElapsedTime->GetValue(unit);
}

File: src/java/com/kitware/pulse/cdm/patient/nutrition/SEMeal.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package com.kitware.pulse.cdm.patient.nutrition;

import com.kitware.pulse.cdm.bind.PatientNutrition.MealData;
import com.kitware.pulse.cdm.properties.SEScalarTime;

public class SEMeal extends SENutrition
{
  protected SEScalarTime  elapsedTime;
  
  public SEMeal()
  {    
    super();
    this.elapsedTime=null;
  }
  
  @Override
  public void clear()
  {
    super.clear();
    if (elapsedTime != null)
      elapsedTime.invalidate();  
  }
  
  public void copy(SEMeal from)
  {
    super.copy(from);
    if(from.hasElapsedTime())
      this.getElapsedTime().set(from.getElapsedTime());
  }
  
  public static void load(MealData src, SEMeal dst)
  {
    dst.clear();
    if(src.hasNutrition())
      SENutrition.load(src.getNutrition(),dst);
    if (src.hasElapsedTime())
      SEScalarTime.load(src.getElapsedTime(),dst.getElapsedTime());
  }
  
  public static MealData unload(SEMeal src)
  {
    MealData.Builder dst = MealData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SEMeal src, MealData.Builder dst)
  {
    SENutrition.unload(src,dst.getNutritionBuilder());
    if (src.hasElapsedTime())
      dst.setElapsedTime(SEScalarTime.unload(src.elapsedTime));
  }    
  
  public SEScalarTime getElapsedTime()
  {
    if (elapsedTime == null)
      elapsedTime = new SEScalarTime();
    return elapsedTime;
  }
  public boolean hasElapsedTime()
  {
    return elapsedTime == null ? false : elapsedTime.isValid();
  }
}
